# 1. Dataframe scraped from Wikipedia

### Scraped from https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

In [16]:
import numpy as np
import pandas as pd
import requests
from bs4 import BeautifulSoup

src = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

In [17]:
soup = BeautifulSoup(src, 'html5lib')

In [31]:
table=soup.find('table')
column_names=['PostalCode','Borough','Neighborhood']
df = pd.DataFrame(columns=column_names)

In [32]:
for tr_cell in table.find_all('tr'):
    row_data=[]
    for td_cell in tr_cell.find_all('td'):
        row_data.append(td_cell.text.strip())
    if len(row_data)==3:
        df.loc[len(df)] = row_data

In [33]:
df.head()

PostalCode           Borough                Neighborhood
0        M1A      Not assigned                            
1        M2A      Not assigned                            
2        M3A        North York                   Parkwoods
3        M4A        North York            Victoria Village
4        M5A  Downtown Toronto  Regent Park / Harbourfront

Below, I found the indexes where "Borough" is "Not assigned". I deleted these row indexes from the dataframe

In [34]:
indexNames = df[ df['Borough'] =='Not assigned'].index
df.drop(indexNames , inplace=True)

df.head(10)

PostalCode           Borough                                  Neighborhood
2         M3A        North York                                     Parkwoods
3         M4A        North York                              Victoria Village
4         M5A  Downtown Toronto                    Regent Park / Harbourfront
5         M6A        North York             Lawrence Manor / Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government
8         M9A         Etobicoke                              Islington Avenue
9         M1B       Scarborough                               Malvern / Rouge
11        M3B        North York                                     Don Mills
12        M4B         East York              Parkview Hill / Woodbine Gardens
13        M5B  Downtown Toronto                      Garden District, Ryerson

Below, I reset the indexes in the dataframe so that they started at 0

In [35]:
result = df.groupby(['PostalCode','Borough'], sort=False).agg( ', '.join)
df_new=result.reset_index()
df_new.head(15)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   
12        M1C       Scarborough   
13        M3C        North York   
14        M4C         East York   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                          Regent Park / Harbourfront  
3                   Lawrence Manor / Lawrence Heights  
4        Queen's Park / Ontario Provincial Government  
5                                    Islington Avenue  
6                                     Malvern / Rouge  
7                                           Don Mills  
8                    Parkview Hill / Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park / Princess Gardens / Martin Gr...  
12           Rouge Hill / Port Union / Highland Creek  
13                                          Don Mills  
14                                   Woodbine Heights

In [23]:
df_new.shape #to find number of rows

(103, 3)

Therefore, there are 103 rows in the dataframe

# 2.  Adding coordinates to neighborhoods

In [24]:
def get_geocode(postal_code):
    # initialize your variable to None
    lat_lng_coords = None
    while(lat_lng_coords is None):
        g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
        lat_lng_coords = g.latlng
    latitude = lat_lng_coords[0]
    longitude = lat_lng_coords[1]
    return latitude,longitude

In [36]:
geo_df=pd.read_csv('http://cocl.us/Geospatial_data')
geo_df.columns=['PostalCode','Latitude','Longitude']
geo_df.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

In [37]:
toronto_df_coors = pd.merge(df_new,
                 geo_df[['PostalCode','Latitude', 'Longitude']],
                 on='PostalCode')
toronto_df_coors

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
23         M4G         East York   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarborough   
27         M2H        North York   
28         M3H        North York   
29         M4H         East York   
..         ...               ...   
73         M4R   Central Toronto   
74         M5R   Central Toronto   
75         M6R      West Toronto   
76         M7R       Mississauga   
77         M9R         Etobicoke   
78         M1S       Scarborough   
79         M4S   Central Toronto   
80         M5S  Downtown Toronto   
81         M6S      West Toronto   
82         M1T       Scarborough   
83         M4T   Central Toronto   
84         M5T  Downtown Toronto   
85         M1V       Scarborough   
86         M4V   Central Toronto   
87         M5V  Downtown Toronto   
88         M8V         Etobicoke   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
92         M5W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
95         M1X       Scarborough   
96         M4X  Downtown Toronto   
97         M5X  Downtown Toronto   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                           Regent Park / Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor / Lawrence Heights  43.718518 -79.464763  
4         Queen's Park / Ontario Provincial Government  43.662301 -79.389494  
5                                     Islington Avenue  43.667856 -79.532242  
6                                      Malvern / Rouge  43.806686 -79.194353  
7                                            Don Mills  43.745906 -79.352188  
8                     Parkview Hill / Woodbine Gardens  43.706397 -79.309937  
9                             Garden District, Ryerson  43.657162 -79.378937  
10                                           Glencairn  43.709577 -79.445073  
11   West Deane Park / Princess Gardens / Martin Gr...  43.650943 -79.554724  
12            Rouge Hill / Port Union / Highland Creek  43.784535 -79.160497  
13                                           Don Mills  43.725900 -79.340923  
14                                    Woodbine Heights  43.695344 -79.318389  
15                                      St. James Town  43.651494 -79.375418  
16                                  Humewood-Cedarvale  43.693781 -79.428191  
17   Eringate / Bloordale Gardens / Old Burnhamthor...  43.643515 -79.577201  
18                 Guildwood / Morningside / West Hill  43.763573 -79.188711  
19                                         The Beaches  43.676357 -79.293031  
20                                         Berczy Park  43.644771 -79.373306  
21                               

# 3. Clustering neighborhoods in Toronto

In [12]:
from geopy.geocoders import Nominatim
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="tl-toronto-neigh")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('Coordinates of Toronto are {}, {}.'.format(latitude, longitude))

Coordinates of Toronto are 43.6534817, -79.3839347.


In [13]:
!conda install -c conda-forge folium=0.5.0 --yes
import folium

Solving environment: done

# All requested packages already installed.



Folium is used to create a Toronto map with bmarkers in blue for each borough

In [38]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

for lat, long, post, borough, neigh in zip(toronto_df_coors['Latitude'], toronto_df_coors['Longitude'], toronto_df_coors['PostalCode'], toronto_df_coors['Borough'], toronto_df_coors['Neighborhood']):
    label = "{} ({}): {}".format(borough, post, neigh)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=popup,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

Reduce boroughs explored to only the ones using Toronto in their name

In [39]:
toronto_boroughs = ['East Toronto', 'Central Toronto', 'Downtown Toronto', 'West Toronto']
toronto_central_df = toronto_df_coors[toronto_df_coors['Borough'].isin(toronto_boroughs)].reset_index(drop=True)
print(toronto_central_df.shape)
toronto_central_df.head()

(39, 5)


PostalCode           Borough                                  Neighborhood  \
0        M5A  Downtown Toronto                    Regent Park / Harbourfront   
1        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government   
2        M5B  Downtown Toronto                      Garden District, Ryerson   
3        M5C  Downtown Toronto                                St. James Town   
4        M4E      East Toronto                                   The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

In [40]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

for lat, long, post, borough, neigh in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['PostalCode'], toronto_central_df['Borough'], toronto_central_df['Neighborhood']):
    label = "{} ({}): {}".format(borough, post, neigh)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=popup,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

In [44]:
CLIENT_ID = 'ZP11JCQSNDEVHZIZP0G2OFAKRUCPI30ER52FZFQOTL2NDDVO' # your Foursquare ID
CLIENT_SECRET = 'ITJ1S2OMINYIZUVZQNVNU4IUZR4YM5PO3JTLG4ZUVJHHKFAG' # your Foursquare Secret
VERSION = '20200426' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: ZP11JCQSNDEVHZIZP0G2OFAKRUCPI30ER52FZFQOTL2NDDVO
CLIENT_SECRET:ITJ1S2OMINYIZUVZQNVNU4IUZR4YM5PO3JTLG4ZUVJHHKFAG


Foursqaure API used to explore the boroughs

In [45]:
radius = 500
LIMIT = 100

venues = []

for lat, long, post, borough, neighborhood in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['PostalCode'], toronto_central_df['Borough'], toronto_central_df['Neighborhood']):
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    for venue in results:
        venues.append((
            post, 
            borough,
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

In [ ]:
venues_df = pd.DataFrame(venues)
venues_df.columns = ['PostalCode', 'Borough', 'Neighborhood', 'BoroughLatitude', 'BoroughLongitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']
print(venues_df.shape)
venues_df.head()

Venues for each postal code

In [47]:
venues_df.groupby(['PostalCode', 'Borough', 'Neighborhood'])['VenueName'].count()


PostalCode  Borough           Neighborhood                                                                                                    
M4E         East Toronto      The Beaches                                                                                                           5
M4K         East Toronto      The Danforth West / Riverdale                                                                                        43
M4L         East Toronto      India Bazaar / The Beaches West                                                                                      23
M4M         East Toronto      Studio District                                                                                                      40
M4N         Central Toronto   Lawrence Park                                                                                                         3
M4P         Central Toronto   Davisville North                                                             

Number of venue categories:

In [48]:
len(venues_df['VenueCategory'].unique())

226

Venues analyzed by area:

In [49]:
# one hot encoding
toronto_central_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

# add postal, borough and neighborhood column back to dataframe
toronto_central_onehot['PostalCode'] = venues_df['PostalCode'] 
toronto_central_onehot['Borough'] = venues_df['Borough'] 
toronto_central_onehot['Neighborhoods'] = venues_df['Neighborhood'] 

# move postal, borough and neighborhood column to the first column
fixed_columns = list(toronto_central_onehot.columns[-3:]) + list(toronto_central_onehot.columns[:-3])
toronto_central_onehot = toronto_central_onehot[fixed_columns]

print(toronto_central_onehot.shape)
toronto_central_onehot.head()

(1623, 229)


PostalCode           Borough               Neighborhoods  Airport  \
0        M5A  Downtown Toronto  Regent Park / Harbourfront        0   
1        M5A  Downtown Toronto  Regent Park / Harbourfront        0   
2        M5A  Downtown Toronto  Regent Park / Harbourfront        0   
3        M5A  Downtown Toronto  Regent Park / Harbourfront        0   
4        M5A  Downtown Toronto  Regent Park / Harbourfront        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Theme Restaurant  \
0                 0                    0  ...                 0   
1                 0                    0  ...                 0   
2                 0                    0  ...                 0   
3                 0                    0  ...                 0   
4                 0                    0  ...                 0   

   Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0                 0      0              0                              0   
1                 0      0              0                              0   
2                 0      0              0                              0   
3                 0      0              0                              0   
4                 0      0              0                              0   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Women's Store  \
0                 0                      0         0              0   
1                 0                      0         0              0   
2                 0                      0         0              0   
3                 0                      0         0              0   
4                 0                      0         0              0   

   Yoga Studio  
0            0  
1            0  
2            0  
3            0  
4            0  

[5 rows x 229 columns]

Frequency of occurance for each category in a specific area

In [50]:
toronto_central_venues_freq = toronto_central_onehot.groupby(['PostalCode', 'Borough', 'Neighborhoods']).mean().reset_index()
print(toronto_central_venues_freq.shape)
toronto_central_venues_freq.head()

(39, 229)


PostalCode          Borough                    Neighborhoods  Airport  \
0        M4E     East Toronto                      The Beaches      0.0   
1        M4K     East Toronto    The Danforth West / Riverdale      0.0   
2        M4L     East Toronto  India Bazaar / The Beaches West      0.0   
3        M4M     East Toronto                  Studio District      0.0   
4        M4N  Central Toronto                    Lawrence Park      0.0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                 0.0           0.0             0.0              0.0   
1                 0.0           0.0             0.0              0.0   
2                 0.0           0.0             0.0              0.0   
3                 0.0           0.0             0.0              0.0   
4                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  ...  Theme Restaurant  \
0               0.0             0.000000  ...               0.0   
1               0.0             0.023256  ...               0.0   
2               0.0             0.000000  ...               0.0   
3               0.0             0.050000  ...               0.0   
4               0.0             0.000000  ...               0.0   

   Toy / Game Store     Trail  Train Station  Vegetarian / Vegan Restaurant  \
0               0.0  0.200000            0.0                            0.0   
1               0.0  0.023256            0.0                            0.0   
2               0.0  0.000000            0.0                            0.0   
3               0.0  0.000000            0.0                            0.0   
4               0.0  0.000000            0.0                            0.0   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Women's Store  \
0               0.0                    0.0     0.000            0.0   
1               0.0                    0.0     0.000            0.0   
2               0.0                    0.0     0.000            0.0   
3               0.0                    0.0     0.025            0.0   
4               0.0                    0.0     0.000            0.0   

   Yoga Studio  
0     0.000000  
1     0.023256  
2     0.000000  
3     0.025000  
4     0.000000  

[5 rows x 229 columns]

10 Most occurred venue types in each area:

In [51]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
areaColumns = ['PostalCode', 'Borough', 'Neighborhoods']
freqColumns = []
for ind in np.arange(num_top_venues):
    try:
        freqColumns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        freqColumns.append('{}th Most Common Venue'.format(ind+1))
columns = areaColumns+freqColumns
# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['PostalCode'] = toronto_central_venues_freq['PostalCode']
neighborhoods_venues_sorted['Borough'] = toronto_central_venues_freq['Borough']
neighborhoods_venues_sorted['Neighborhoods'] = toronto_central_venues_freq['Neighborhoods']

for ind in np.arange(toronto_central_venues_freq.shape[0]):
    row_categories = toronto_central_venues_freq.iloc[ind, :].iloc[3:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    neighborhoods_venues_sorted.iloc[ind, 3:] = row_categories_sorted.index.values[0:num_top_venues]

neighborhoods_venues_sorted.sort_values(freqColumns, inplace=True)
neighborhoods_venues_sorted

PostalCode           Borough  \
27        M5V  Downtown Toronto   
31        M6H      West Toronto   
32        M6J      West Toronto   
25        M5S  Downtown Toronto   
3         M4M      East Toronto   
15        M5C  Downtown Toronto   
26        M5T  Downtown Toronto   
33        M6K      West Toronto   
24        M5R   Central Toronto   
14        M5B  Downtown Toronto   
6         M4R   Central Toronto   
19        M5J  Downtown Toronto   
29        M5X  Downtown Toronto   
18        M5H  Downtown Toronto   
16        M5E  Downtown Toronto   
20        M5K  Downtown Toronto   
17        M5G  Downtown Toronto   
28        M5W  Downtown Toronto   
13        M5A  Downtown Toronto   
9         M4V   Central Toronto   
21        M5L  Downtown Toronto   
37        M7A  Downtown Toronto   
5         M4P   Central Toronto   
7         M4S   Central Toronto   
2         M4L      East Toronto   
22        M5N   Central Toronto   
35        M6R      West Toronto   
1         M4K      East Toronto   
30        M6G  Downtown Toronto   
23        M5P   Central Toronto   
38        M7Y      East Toronto   
34        M6P      West Toronto   
4         M4N   Central Toronto   
10        M4W  Downtown Toronto   
36        M6S      West Toronto   
8         M4T   Central Toronto   
0         M4E      East Toronto   
11        M4X  Downtown Toronto   
12        M4Y  Downtown Toronto   

                                        Neighborhoods 1st Most Common Venue  \
27  CN Tower / King and Spadina / Railway Lands / ...       Airport Service   
31                      Dufferin / Dovercourt Village                Bakery   
32                          Little Portugal / Trinity                   Bar   
25                    University of Toronto / Harbord                  Café   
3                                     Studio District                  Café   
15                                     St. James Town                  Café   
26        Kensington Market / Chinatown / Grange Park                  Café   
33     Brockton / Parkdale Village / Exhibition Place                  Café   
24              The Annex / North Midtown / Yorkville                  Café   
14                           Garden District, Ryerson        Clothing Store   
6                                  North Toronto West        Clothing Store   
19  Harbourfront East / Union Station / Toronto Is...           Coffee Shop   
29            First Canadian Place / Underground city           Coffee Shop   
18                         Richmond / Adelaide / King           Coffee Shop   
16                                        Berczy Park           Coffee Shop   
20          Toronto Dominion Centre / Design Exchange           Coffee Shop   
17                                 Central Bay Street           Coffee Shop   
28                                     Stn A PO Boxes           Coffee Shop   
13                         Regent Park / Harbourfront           Coffee Shop   
9   Summerhill West / Rathnelly / South Hill / For...           Coffee Shop   
21                    Commerce Court / Victoria Hotel           Coffee Shop   
37       Queen's Park / Ontario Provincial Government           Coffee Shop   
5                                    Davisville North          Dance Studio   
7                                          Davisville          Dessert Shop   
2                     India Bazaar / The Beaches West  Fast Food Restaurant   
22                                           Roselawn                Garden   
35                            Parkdale / Roncesvalles             Gift Shop   
1                       The Danforth West / Riverdale      Greek Restaurant   
30                                           Christie         Grocery Store   
23                           Forest Hill North & West         Jewelry Store   
38              Business reply mail Processing CentrE    Light Rail Station   
34                     High Park / The Junction South    Mexican Restauran

KMeans algorithm used to cluster Toronto Central areas into 3 clusters (k=3)

In [53]:
from sklearn.cluster import KMeans

kclusters = 3

toronto_central_venues_freq_clustering = toronto_central_venues_freq.drop(['PostalCode', 'Borough', 'Neighborhoods'], 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_central_venues_freq_clustering)

toronto_central_clustered_df = toronto_central_df
toronto_central_clustered_df['Cluster'] = kmeans.labels_

toronto_central_clustered_df = toronto_central_clustered_df.join(neighborhoods_venues_sorted.drop(['Borough', 'Neighborhoods'], 1).set_index('PostalCode'), on='PostalCode')
toronto_central_clustered_df.sort_values(['Cluster'] + freqColumns, inplace=True)
toronto_central_clustered_df

PostalCode           Borough  \
32        M5V  Downtown Toronto   
9         M6H      West Toronto   
11        M6J      West Toronto   
27        M5S  Downtown Toronto   
17        M4M      East Toronto   
3         M5C  Downtown Toronto   
30        M5T  Downtown Toronto   
14        M6K      West Toronto   
24        M5R   Central Toronto   
2         M5B  Downtown Toronto   
23        M4R   Central Toronto   
10        M5J  Downtown Toronto   
36        M5X  Downtown Toronto   
5         M5E  Downtown Toronto   
13        M5K  Downtown Toronto   
6         M5G  Downtown Toronto   
34        M5W  Downtown Toronto   
0         M5A  Downtown Toronto   
31        M4V   Central Toronto   
16        M5L  Downtown Toronto   
1         M7A  Downtown Toronto   
20        M4P   Central Toronto   
26        M4S   Central Toronto   
15        M4L      East Toronto   
19        M5N   Central Toronto   
25        M6R      West Toronto   
12        M4K      East Toronto   
7         M6G  Downtown Toronto   
21        M5P   Central Toronto   
38        M7Y      East Toronto   
18        M4N   Central Toronto   
33        M4W  Downtown Toronto   
28        M6S      West Toronto   
29        M4T   Central Toronto   
4         M4E      East Toronto   
35        M4X  Downtown Toronto   
37        M4Y  Downtown Toronto   
8         M5H  Downtown Toronto   
22        M6P      West Toronto   

                                         Neighborhood   Latitude  Longitude  \
32  CN Tower / King and Spadina / Railway Lands / ...  43.628947 -79.394420   
9                       Dufferin / Dovercourt Village  43.669005 -79.442259   
11                          Little Portugal / Trinity  43.647927 -79.419750   
27                    University of Toronto / Harbord  43.662696 -79.400049   
17                                    Studio District  43.659526 -79.340923   
3                                      St. James Town  43.651494 -79.375418   
30        Kensington Market / Chinatown / Grange Park  43.653206 -79.400049   
14     Brockton / Parkdale Village / Exhibition Place  43.636847 -79.428191   
24              The Annex / North Midtown / Yorkville  43.672710 -79.405678   
2                            Garden District, Ryerson  43.657162 -79.378937   
23                                 North Toronto West  43.715383 -79.405678   
10  Harbourfront East / Union Station / Toronto Is...  43.640816 -79.381752   
36            First Canadian Place / Underground city  43.648429 -79.382280   
5                                         Berczy Park  43.644771 -79.373306   
13          Toronto Dominion Centre / Design Exchange  43.647177 -79.381576   
6                                  Central Bay Street  43.657952 -79.387383   
34                                     Stn A PO Boxes  43.646435 -79.374846   
0                          Regent Park / Harbourfront  43.654260 -79.360636   
31  Summerhill West / Rathnelly / South Hill / For...  43.686412 -79.400049   
16                    Commerce Court / Victoria Hotel  43.648198 -79.379817   
1        Queen's Park / Ontario Provincial Government  43.662301 -79.389494   
20                                   Davisville North  43.712751 -79.390197   
26                                         Davisville  43.704324 -79.388790   
15                    India Bazaar / The Beaches West  43.668999 -79.315572   
19                                           Roselawn  43.711695 -79.416936   
25                            Parkdale / Roncesvalles  43.648960 -79.456325   
12                      The Danforth West / Riverdale  43.679557 -79.352188   
7                                            Christie  43.669542 -79.422564   
21                           Forest Hill North & West  43.696948 -79.411307   
38              Business reply mail Processing CentrE  43.662744 -79.321558   
18                                      Lawrence Park  43.728020 -79.388790   
33                                           Rosedale  43.679563 -79.37752

The clusters are shown on the map

In [56]:
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, post, bor, poi, cluster in zip(toronto_central_clustered_df['Latitude'], toronto_central_clustered_df['Longitude'], toronto_central_clustered_df['PostalCode'], toronto_central_clustered_df['Borough'], toronto_central_clustered_df['Neighborhood'], toronto_central_clustered_df['Cluster']):
    label = folium.Popup('{} ({}): {} - Cluster {}'.format(bor, post, poi, cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters